### PR TITLE
Add service monitor for operand

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

File: bindata/v3.11.0/kube-apiserver/prometheus-service-monitor.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+metadata:
+  name: cluster-kube-apiserver
+  namespace: openshift-kube-apiserver
+spec:
+  endpoints:
+    - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+      interval: 30s
+      metricRelabelings:
+        - action: drop
+          regex: etcd_(debugging|disk|request|server).*
```
Comment (sttts) [inline bindata/v3.11.0/kube-apiserver/prometheus-service-monitor.yaml]: where does this file come from?

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+var monitorGroupVersionKind = schema.GroupVersionResource{Group: "monitoring.coreos.com", Version: "v1", Resource: "servicemonitors"}
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: monitorGVR, not kind

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+// ApplyServiceMonitor applies the Prometheus service monitor.
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: what does "requires rules" mean?

Comment (mfojtik) [reply vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: fixed, bad copy & paste.

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+func ApplyServiceMonitor(client dynamic.Interface, recorder events.Recorder, serviceMonitorBytes []byte) (bool, error) {
+	monitorJSON, err := yaml.YAMLToJSON(serviceMonitorBytes)
+	if err != nil {
+		panic(err)
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: return err

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	monitorObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, monitorJSON)
+	if err != nil {
+		panic(err)
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: return err

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	required, ok := monitorObj.(*unstructured.Unstructured)
+	if !ok {
+		panic(fmt.Sprintf("unexpected object in %t", monitorObj))
+	}
+
+	existing, err := client.Resource(monitorGroupVersionKind).Get(required.GetName(), metav1.GetOptions{})
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: no namespace?

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	monitorObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, monitorJSON)
+	if err != nil {
+		panic(err)
+	}
+	required, ok := monitorObj.(*unstructured.Unstructured)
+	if !ok {
+		panic(fmt.Sprintf("unexpected object in %t", monitorObj))
+	}
+
+	existing, err := client.Resource(serviceMonitorGVR).Get(required.GetName(), metav1.GetOptions{})
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: namespace

File: pkg/operator/target_config_reconciler_v311_00.go [hunk 8]
```diff
@@ -49,6 +51,10 @@ func createTargetConfigReconciler_v311_00_to_latest(c TargetConfigReconciler, re
 	if err != nil {
 		errors = append(errors, fmt.Errorf("%q: %v", "configmap/kube-apiserver-pod", err))
 	}
+	_, err = managePrometheus_v311_00_to_latest(c.kubeClient.RbacV1(), c.dynamicClient, recorder)
+	if err != nil {
```
Comment (sttts) [inline pkg/operator/target_config_reconciler_v311_00.go]: `if _, err := ... ; err != nil `

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
+package resourceapply
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/library-go/pkg/operator/events"
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: order

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	if !ok {
+		panic(fmt.Sprintf("unexpected object in %t", monitorObj))
+	}
+
+	existing, err := client.Resource(serviceMonitorGVR).Get(required.GetName(), metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		_, createErr := client.Resource(serviceMonitorGVR).Create(required)
+		if createErr != nil {
+			recorder.Warningf("ServiceMonitorCreateFailed", "Failed to create ServiceMonitor.monitoring.coreos.com/v1: %v", createErr)
+			return true, createErr
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: not `return false, createErr`?

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+	if equality.Semantic.DeepEqual(existingSpec, requiredSpec) {
+		return false, nil
+	}
+
+	if err := unstructured.SetNestedField(existing.UnstructuredContent(), requiredSpec, "spec"); err != nil {
+		return true, err
+	}
+
+	_, err = client.Resource(serviceMonitorGVR).Update(existing)
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: namespace

if _, err := ...; err != nil

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go (hunk 10 above)
Comment (mfojtik) [reply vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: no, we return true in this case from other apply func... i guess it indicates that we detected a change (not found) but failed to react to it (error)

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	if err == nil {
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: !=

File: vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+		recorder.Warningf("ServiceMonitorUpdateFailed", "Failed to update ServiceMonitor.monitoring.coreos.com/v1: %v", err)
+	} else {
+		recorder.Eventf("ServiceMonitorUpdated", "Updated ServiceMonitor.monitoring.coreos.com/v1 because it changed")
+	}
+
+	return true, err
```
Comment (sttts) [inline vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: true only in `err == nil` case

Comment (mfojtik) [reply vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go]: i think same logic as create applies here (we found a change, we failed to react (so true, error))

**Comment by mfojtik** [comment]
@sttts thanks for review! addressed all comments except 2 ;-)

@deads2k if you're happy with this (and tests will pass), I will do real bump in library-go and make similar change in openshift-apiserver.

I'm not sure about controller manager, I think we postponed serving metrics/etc. as we can't serve them securely until we get that in 1.12 rebase?

File: bindata/v3.11.0/kube-apiserver/svc.yaml [hunk 14]
```diff
@@ -5,12 +5,10 @@ metadata:
   name: apiserver
   annotations:
     service.alpha.openshift.io/serving-cert-secret-name: serving-cert
-    prometheus.io/scrape: "true"
-    prometheus.io/scheme: https
 spec:
   type: ClusterIP
   selector:
-    apiserver: "true"
```
[the remark this answers is not in the document]
Comment (deads2k) [reply bindata/v3.11.0/kube-apiserver/svc.yaml]: > do we use this label for anything? if not I would drop it from the pod maybe?

sure

File: pkg/operator/target_config_reconciler_v311_00.go [hunk 15]
```diff
@@ -50,6 +52,10 @@ func createTargetConfigReconciler_v311_00_to_latest(c TargetConfigReconciler, re
 		errors = append(errors, fmt.Errorf("%q: %v", "configmap/kube-apiserver-pod", err))
 	}
 
+	if _, err = managePrometheus_v311_00_to_latest(c.kubeClient.RbacV1(), c.dynamicClient, recorder); err != nil {
```
Comment (deads2k) [inline pkg/operator/target_config_reconciler_v311_00.go]: model all this as a separate control loop.  This loop shouldn't trigger a failing status and is logically generic for all of our operators.

Comment (mfojtik) [reply pkg/operator/target_config_reconciler_v311_00.go]: i think it makes sense to make this generic. https://github.com/openshift/library-go/pull/150

**Comment by mfojtik** [comment]
replaced by https://github.com/openshift/library-go/pull/150

**Pull request closed without merging**